### ISSUE TITLE
DSP Overflow - high pixel values are being clamped when running on DSP

### ISSUE THREAD

**Issue by aviaisr** [tensorflow/tensorflow#53552]
Hi,
I trained a keras model to extract gray-level segmentation maps. 
I converted the model to TFLite and quantized the model.
The quantized model produces similar results on CPU and DSP HWs, if the values of the pixels are not very high (<<1).
If the maps produce pixel predictions with high values that are closer to 1, it seems that the values are being clamped when running on DSP, unlike in CPU which produces reasonable maps.

I repeated the tests on versions 2.2, 2.4, and 2.7 and all demonstrated the same results.

What can I do to change this?
Should I change the dynamic range of the maps?

**Comment by tilakrayal** [comment]
@aviaisr ,
In order to expedite the trouble-shooting process, could you please provide the complete code and dataset to reproduce the issue reported here.

**Comment by aviaisr** [comment]
1. I saved the kears model + tflite file + quantized tflite file here: https://drive.google.com/drive/folders/1ZIsPCiBPCzDmFcTbVAokjM7CtzLJO8y7?usp=sharing

2. Conversion code
_# Quntization script

# read an hdf5 keras model and convert to tflite with quantization
import tensorflow as tf
import numpy as np
import cv2
import sys
import os

def rep_data_gen0():
    a = []
    for i in range(128):
        img = 5*np.fromfile('norm_images/image_'+str(i)+".bin", dtype=np.float32)
        img = img.reshape(224, 224, 3)
        a.append(img)

    a = np.array(a)
    print(a.shape)
    img = tf.data.Dataset.from_tensor_slices(a).batch(1)
    for i in img.take(128):
        yield [i]

def convert_and_quantize(model):
    converter = tf.lite.TFLiteConverter.from_keras_model(model)
    converter.optimizations = [tf.lite.Optimize.DEFAULT]
    converter.representative_dataset = rep_data_gen0
    converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS_INT8]

    print("Converting to TfLite using full integer quantization")
    quant_model = converter.convert()
    return quant_model

def main_func():
    print (f'Reading keras model from {sys.argv[1]}')
    keras_model = tf.keras.models.load_model(sys.argv[1], compile = False)
    input_name = keras_model.input_names[0]
    index = keras_model.input_names.index(input_name)
    keras_model.inputs[index].set_shape([1, 224, 224, 3])
    keras_model.trainable = False
    quant_model = convert_and_quantize(keras_model)

    print (f'Writing quantized tflite model to {sys.argv[2]}')
    f = open(sys.argv[2],'wb')
    f.write(quant_model)
    f.close()
    

if __name__ == "__main__":
    print(tf.__version__)
    if len(sys.argv) != 3:
        print('Wrong command line arguments')
        sys.exit(1)
    main_func()
    print("Done")_

3. Example dataset (I train on 5M images, so I just uploaded 400 images as example). My task is a segmentation task and the trained maps are binary maps.
https://drive.google.com/drive/folders/1Dn0olSmZ9HFGwAIViq1cfl9kfuvC5Qs_?usp=sharing

4. representative dataset as bins saved in 
https://drive.google.com/drive/folders/1Q0iqM1qHdWFdW_PyA0BEQtzde9dmU7S5?usp=sharing

5. Main training code details:

_trainDataset = tf.data.Dataset.from_tensor_slices(dataPaths['train']) #same for validation
batchSize = 32

model = Unet('mobilenetv2', input_shape=inputShape, encoder_weights='imagenet', encoder_freeze=False,
              decoder_block_type='transpose', decoder_filters=decoderFilters)  # Deconvolution
              
model = utils.set_regularization(model, kernel_regularizer=keras.regularizers.l2(0.001),
                                  bias_regularizer=keras.regularizers.l2(0.001))

lr = keras.callbacks.LearningRateScheduler(scheduler)

loss = losses.binary_focal_dice_loss

checkpoint = keras.callbacks.ModelCheckpoint(weightsPath, monitor='val_loss', verbose=0,
                                              save_best_only=True, save_weights_only=False,
                                              mode='auto', save_freq='epoch')

callbacks_list = [lr, checkpoint, historyLogger]

# compile
adam = keras.optimizers.Adam(lr=0.01, beta_1=0.9, beta_2=0.999, epsilon=1e-07, decay=0.0, amsgrad=False)

model.compile(loss=loss, optimizer=adam)

history = model.fit(x=trainDataset,
                    batch_size=batchSize,
                    epochs=numEpochs,
                    verbose=1,  # show progress bar
                    validation_data=valDataset,
                    initial_epoch=initialEpoch,
                    steps_per_epoch=stepsPerEpoch,
                    validation_steps=validationSteps,
                    callbacks=callbacks_list)_

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
